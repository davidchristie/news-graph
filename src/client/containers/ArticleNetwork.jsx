import React, { PropTypes } from 'react'
import { connect } from 'react-redux'
import vis from 'vis'

const options = {
  nodes: {
    borderWidth: 4,
    size: 30,
    color: {
      border: '#406897',
      background: '#6AAFFF'
    },
    font: {color: '#eeeeee'},
    shape: 'dot'
  },
  edges: {
    arrows: 'to',
    color: 'lightgray'
  }
}

export const ArticleNetwork = function ArticleNetwork (props) {
  const articles = props.articles
  const connections = props.connections
  const data = {articles, connections}
  const style = {
    width: '100%',
    height: '400px',
    border: '1px solid lightgray',
    backgroundColor: '#333333'
  }
  return (
    <div
      ref={
        container => {
          if (container) {
            createNetwork(container, data, options)
          }
        }}
      style={style}
    />
  )
}

ArticleNetwork.propTypes = {
  article: PropTypes.shape({
    id: PropTypes.number.isRequired
  }),
  articles: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number.isRequired,
    title: PropTypes.string.isRequired
  }).isRequired).isRequired,
  connections: PropTypes.arrayOf(PropTypes.shape({
    from: PropTypes.number.isRequired,
    id: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired,
    to: PropTypes.number.isRequired
  }).isRequired).isRequired
}

function articleToNode (article) {
  return {
    id: article.id,
    label: article.title
  }
}

function connectionToEdge (connection) {
  return {
    from: connection.from,
    id: connection.id,
    label: connection.name,
    to: connection.to
  }
}

function createNetwork (container, data, options) {
  const { articles, connections } = data
  const nodes = articles.map(articleToNode)
  const edges = connections.map(connectionToEdge)
  return new vis.Network(container, {edges, nodes}, options)
}

function stateToProps (state) {
  return {
    articles: state.articles,
    connections: state.connections
  }
}

export default connect(stateToProps)(ArticleNetwork)
