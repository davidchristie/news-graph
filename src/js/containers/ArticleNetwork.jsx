import React, { PropTypes } from 'react'
import { connect } from 'react-redux'
import vis from 'vis'

const ArticleNetwork = class ArticleNetwork extends React.Component {
  constructor(props) {
    super(props)
    this.createNetwork = this.createNetwork.bind(this)
  }
  createNetwork() {

    const articles = this.props.articles

    const nodes = articles.map(article => {
      return {
        id: article.id,
        label: article.url
      }
    })

    const edges = []

    const data = {
      nodes,
      edges
    }

    const options = {
      nodes: {
        borderWidth:4,
        size:30,
        color: {
          border: '#406897',
          background: '#6AAFFF'
        },
        font:{color:'#eeeeee'},
        shape: 'dot'
      },
      edges: {
        color: 'lightgray'
      }
    }

    new vis.Network(this.container, data, options)
  }
  render() {
    return (
      <div
        ref={container => {
          if (container) {
            this.container = container
            this.createNetwork()
          }
        }}
        style={{
          width: '100%',
          height: '400px',
          border: '1px solid lightgray',
          backgroundColor: '#333333'
        }}
      />
    )
  }
}

ArticleNetwork.propTypes = {
  articles: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number.isRequired,
    url: PropTypes.string.isRequired
  }).isRequired).isRequired
}

function mapStateToProps(state) {
  return {
    articles: state.app.articles
  }
}

export default connect(mapStateToProps)(ArticleNetwork)
