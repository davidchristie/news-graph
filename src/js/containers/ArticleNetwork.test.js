/* eslint-env jest */

import React from 'react'
import { Provider } from 'react-redux'
import renderer from 'react-test-renderer'
import configureStore from 'redux-mock-store'

import ArticleNetwork from './ArticleNetwork'

const initialState = {
  app: {
    articles: [
      {
        id: 1,
        url: 'http://www.nature.com/news/ancient-retroviruses-emerged-half-a-billion-years-ago-1.21274'
      },
      {
        id: 2,
        url: 'https://www.theguardian.com/world/2017/jan/11/chamber-of-rats-mexico-parliaments-name-changed-in-google-maps-prank'
      },
      {
        id: 3,
        url: 'https://vimeo.com/192574852'
      },
      {
        id: 4,
        url: 'http://www.aljazeera.com/indepth/features/2017/01/india-love-hand-books-170108073151216.html'
      }
    ],
    connections: [
      {
        from: 1,
        id: 1,
        name: 'Sourced by',
        to: 2
      },
      {
        from: 2,
        id: 2,
        name: 'Sourced by',
        to: 3
      },
      {
        from: 3,
        id: 3,
        name: 'Response to',
        to: 1
      },
      {
        from: 3,
        id: 4,
        name: 'Response to',
        to: 2
      }
    ]
  }
}
const store = configureStore()(initialState)

test('ArticleNetwork renders correctly', () => {
  const tree = renderer.create(
    <Provider store={store}>
      <ArticleNetwork />
    </Provider>
  ).toJSON()
  expect(tree).toMatchSnapshot()
})
