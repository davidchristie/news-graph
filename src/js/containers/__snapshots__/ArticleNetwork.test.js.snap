exports[`test ArticleNetwork renders correctly 1`] = `
<div
  style={
    Object {
      "backgroundColor": "#333333",
      "border": "1px solid lightgray",
      "height": "400px",
      "width": "600px",
    }
  } />
`;
